DAG deploy documentation (#1381)
astronomer/docs: astro/deploy-code.md

After you select a Deployment, the CLI parses your DAGs to ensure that they don't contain basic syntax and import errors. This test is equivalent to the one that runs during `astro dev parse` in a local Airflow environment. If any of your DAGs fail this parse, the deploy to Astro also fails.

If your code passes the parse, the Astro CLI builds your Astro project directory into a new Docker image and then pushes the image to your Deployment on Astro. To force a deploy even if your project has DAG errors, you can run `astro deploy --force`.

:::tip

To validate your code before deploying it to Astro, you can run `astro deploy --pytest`. Adding the `--pytest` flag makes the CLI run all tests in your project's `tests` directory using [pytest](https://docs.pytest.org/en/7.0.x/contents.html). If any of these tests fail, your code deploy also fails. This can help you prevent your team from deploying DAGs to Astro that contain errors.

For more information about using Pytest, see [Test and troubleshoot locally](test-and-troubleshoot-locally.md#test-dags-locally-with-pytest).

:::

### Step 3: Validate your changes

> After the deploy completes, Docker image information for your Deployment is available in the **Image tag** field in the footer of the Airflow UI. Depending on how your organization deploys to Astro, the **Image tag** field displays a unique identifier generated by a Continuous Integration (CI) tool or a timestamp generated by the Astro CLI on `astro deploy`. The **Image tag** field in the Airflow UI footer identifies the Docker image running on the webserver of your Deployment. This information can help you determine if your deploy was successful, but it does not identify the Docker image running on your scheduler, triggerer, or workers. To confirm a code push with complete certainty, contact [Astronomer support](https://cloud.astronomer.io/support).

Author: ```suggestion
After the deploy completes, Docker image information for your Deployment is available in the **Image tag** field in the footer of the Airflow UI. Depending on how your organization deploys to Astro, the **Image tag** field displays a unique identifier generated by a Continuous Integration (CI) tool or a timestamp generated by the Astro CLI on `astro deploy`. The **Image tag** field in the Airflow UI footer identifies the Docker image running on the webserver of your Deployment. This information can help you determine if your deploy was successful, but it does not identify the Docker image running on your scheduler, triggerer, or workers. To fully confirm a code push, contact [Astronomer support](https://cloud.astronomer.io/support).
```

:::

## Step 2: Push your Astro project to an Astro Deployment

To deploy your Astro project, run:

```sh
astro deploy
```

This command returns a list of Deployments available in your Workspace and prompts you to pick one.

After you select a Deployment, the CLI parses your DAGs to ensure that they don't contain basic syntax and import errors. This test is equivalent to the one that runs during `astro dev parse` in a local Airflow environment. If any of your DAGs fail this parse, the deploy to Astro also fails.

> If your code passes the parse, the CLI builds your Astro project directory into a new Docker image and then pushes the image to your Deployment on Astro. To force a deploy even if your project has DAG errors, you can run `astro deploy --force`.

Reviewer: ```suggestion
If your code passes the parse, the Astro CLI builds your Astro project directory into a new Docker image and then pushes the image to your Deployment on Astro. To force a deploy even if your project has DAG errors, you can run `astro deploy --force`.
```

> :::

Author: Let's just point to `## Dag-only deploys` even if we don't have ALL the deets right now instead of promising future documentation, also I don't think this needs to be an info box.

## DAG-only deploys

Running `astro deploy` without additional flags deploys your entire Astro project, including your DAG code, Dockerfile, and dependencies. There are some scenarios where you need to deploy only your Astro project's `dags` directory:

- You work on an Astro project where one team member is responsible for project configuration changes and another team member is responsible for DAG authoring.
- You want to deploy DAGs without building a Docker image or using the Astro CLI. For example, you upload DAGs to a shared S3 bucket.
- You want to deploy DAG changes without downtime to your Deployment. 

In these cases, you can enable DAG-only deploys on a Deployment.

### Enable DAG-only deploys

```sh
astro deployment update --dag-deploy enable
```

> The CLI prompts you to select a Deployment to enable the feature on. After you enable the feature, anyone with access to the Deployment can deploy DAGs.

Reviewer: ```suggestion
When the prompt appears in the Astro CLI, select a Deployment  where you want to enable the feature. After you enable the feature, anyone with access to the Deployment can deploy DAGs.
```

:::caution 

> Enabling DAG deploys pauses all DAGS in a Deployment. Any active DAG runs will finish, but new DAG runs will not be scheduled until you unpause the DAGs from the Deployment's Airflow UI.

Reviewer: ```suggestion
When you enable DAG-only deploys, all DAGS in a Deployment are paused. Any active DAG runs will finish, but new DAG runs will not be scheduled until you unpause the DAGs in the Airflow UI.
```

:::

### Trigger a DAG-only deploy

Run the following command to deploy only your current `dags` directory to a Deployment:

```sh
astro deploy --dags
```

> This command pushes DAGs to a component on Astro that handles loading DAGs into your Airflow environment. DAG only deploys do not disrupt any active task runs, resulting in zero downtime for the Deployment. If you deploy changes to a currently running DAG, any active task runs finish using code from before the Deploy. Any new task runs are scheduled using code from after the deploy. 

Reviewer: ```suggestion
This command pushes DAGs to a component on Astro that loads DAGs into your Airflow environment. DAG only deploys do not disrupt any active task runs, resulting in zero downtime for the Deployment. If you deploy changes to a currently running DAG, any active task runs finish using code from before the Deploy. Any new task runs are scheduled using code from after the deploy. 
```

> DAG runs could fail due to downstream tasks running code from a different source than their upstream tasks. DAG runs that fail this way need to be fully restarted from the Airflow UI so that all tasks are executed based on the same source code.

Reviewer: ```suggestion
DAG runs could fail due to downstream tasks running code from a different source than their upstream tasks. DAG runs that fail this way need to be fully restarted from the Airflow UI so that all tasks are executed with the same source code.
```

:::

## DAG-only deploys

There are some scenarios where you might want to deploy only the `dags` directory of your Astro project. To do this, you can enable the DAG-only deploy feature per Deployment.

Enabling DAG-only deploys on Astro has a few benefits:

- DAG-based deploys are significantly faster than image-only deploys when you deploy only changes to your DAGs. Your updated code will appear in the Airflow UI much faster.
- When you push only changes to your DAGs, the workers and schedulers in your Deployment will pick up your changes gracefully and will not restart. This results in improved reliability, a more efficient use of running workers, and no downtime for your Deployment.
- You can have different sets of users deploy project changes versus DAG changes. See [DAG-based workflows](ci-cd.md#dag-based-workflows) for how you can set this up in your CI/CD pipelines.

### Enable DAG-only deploys

> When the DAG-only deploy feature is enabled, DAGs are deployed separately and are never built into or included in the Docker image for your Astro project. You must still run `astro deploy` when you make a change to any file in your Astro project that is not in the `dags` directory. All files in your Astro project that are not in the `dags` directory are built into a Docker image and pushed to Astro.

Author: ```suggestion
```

To push DAG-only changes to Astro, you must enable the feature for each Deployment. Enabling DAG-only deploys requires Astro CLI version 1.7 or above. See [Install the Astro CLI](cli/install-cli.md).

1. Run:

    ```sh
    astro deployment update --dag-deploy enable
    ```

> 2. When the prompt appears in the Astro CLI, select a Deployment where you want to enable the feature. Running tasks will not be interrupted and new tasks will continue to be scheduled.

Author: ```suggestion
2. When the prompt appears in the Astro CLI, select a Deployment where you want to enable the feature. Running tasks will not be interrupted and new tasks will continue to be scheduled.

When the DAG-only deploy feature is enabled, DAGs are deployed separately and are never built into or included in the Docker image for your Astro project. You must still run `astro deploy` when you make a change to any file in your Astro project that is not in the `dags` directory. All files in your Astro project that are not in the `dags` directory are built into a Docker image and pushed to Astro.
```
